### PR TITLE
package DataFrame for conan-io

### PR DESCRIPTION
**Pull request by yssource**
Specify library name and version:  **DataFrame/1.5.0**

- [ x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [ x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/DataFrame/config.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
```
Comment (danimtb) [inline recipes/DataFrame/config.yml]:
```suggestion
```

File: recipes/DataFrame/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            os.path.join("lib", "pkgconfig"),
+        ]:
+            tools.rmdir(os.path.join(self.package_folder, dir_to_remove))
+
+    def package_info(self):
+        self.cpp_info.libs = tools.collect_libs(self)
+
+        # in linux we need to link also with these libs
+        if self.settings.os == "Linux":
+            self.cpp_info.libs.extend(["pthread", "dl", "rt"])
```
Comment (danimtb) [inline recipes/DataFrame/all/conanfile.py]:
```suggestion
            self.cpp_info.system_libs.extend(["pthread", "dl", "rt"])
```

**Comment by danimtb** [comment]
Don't know why this generated just 1 package 🤔

**Comment by yssource** [comment]
Thanks for reviews.

**Comment by yssource** [comment]
It's necessarily to rebase the commits, isn't it?

**Comment by yssource** [comment]
It seems the failure that is caused under "Windows x86_64". But I have not such guest developing environment on my local PC.

**Comment by uilianries** [comment]
> It seems the failure that is caused under "Windows x86_64". But I don't have such a guest developing environment on my local PC.

You don't need, just check the logs provided by the CI.

The problem is that on Windows, this project copies those CMake file to `self.package_folder/CMake`, instead of `self.package_folder/lib/cmake`, like on Unix systems.

File: recipes/dataframe/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    def package(self):
+        self.copy("License", dst="licenses", src=self._source_subfolder)
+        cmake = self._configure_cmake()
+        cmake.install()
+
+        # Remove packaging files & MS runtime files
+        for dir_to_remove in [
+            os.path.join("lib", "cmake"),
+            os.path.join("lib", "share"),
+            os.path.join("lib", "pkgconfig"),
```
Comment (uilianries) [inline recipes/dataframe/all/conanfile.py]:
```suggestion
            os.path.join("lib", "pkgconfig"),
            "CMake"
```

Comment (uilianries) [reply recipes/dataframe/all/conanfile.py]: This should solve your error on Windows

**Comment by yssource** [comment]
``` python
        # Remove packaging files & MS runtime files
        for dir_to_remove in [
            "cmake",
            os.path.join("lib", "cmake"),
            os.path.join("lib", "share"),
            os.path.join("lib", "pkgconfig"),
            "share"
        ]:
            tools.rmdir(os.path.join(self.package_folder, dir_to_remove))

```

@uilianries  I'm not sure, whether it is ok or not, as above.

**Comment by uilianries** [comment]
@yssource You fixed one error, now you have a new one:

DataFrame requires C++17, however, MSVC 14 and MSVC 15 doesn't offer support for such cpp standard. You need to drop them by configure.

**Comment by uilianries** [comment]
@yssource reading again the author cmake file, it requires cpp 14 actually.

File: recipes/dataframe/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_folder = "DataFrame-V-" + self.version
+        os.rename(extracted_folder, self._source_subfolder)
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
```
Comment (uilianries) [inline recipes/dataframe/all/conanfile.py]:
```suggestion

    def configure(self):
        if self.settings.compiler == "Visual Studio" and \
           Version(self.settings.compiler.version) < "15":
            raise ConanInvalidConfiguration("DataFrame requires Visual Studio >= 15")
```

File: recipes/dataframe/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
+import os
+from conans import ConanFile, CMake, tools
```
Comment (uilianries) [inline recipes/dataframe/all/conanfile.py]:
```suggestion
from conans import ConanFile, CMake, tools
from conans.tools import Version
from conans.errors import ConanInvalidConfiguration
```

File: recipes/dataframe/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    def configure(self):
+        compiler = self.settings.compiler
+        if (
+            compiler == "Visual Studio"
+            and Version(self.settings.compiler.version) < "16"
+        ):
+            raise ConanInvalidConfiguration("DataFrame requires Visual Studio >= 16")
+        elif compiler == "apple-clang":
+            raise ConanInvalidConfiguration("DataFrame is not tested with apple-clang")
```
Comment (uilianries) [inline recipes/dataframe/all/conanfile.py]: What do you mean by "not tested"? Who didn't test, you or the author? You should check if it's working on OSX.

Comment (yssource) [reply recipes/dataframe/all/conanfile.py]: OSX tests are commented out in the official [hosseinmoein/DataFrame/.travis.yml](https://github.com/hosseinmoein/DataFrame/blob/master/.travis.yml).
emm, I will test this in my fork with travis-ci.

**Comment by yssource** [comment]
The building on Visual Studio has troubled me for a long time.
I created a repo with the same codes and made a successful building on [travis-ci](https://travis-ci.com/yssource/conan-center-index-dataframe/builds/144002893)  and [appveyor](https://ci.appveyor.com/project/yssource/conan-center-index-dataframe).
But, it fails to build here.

**Comment by uilianries** [comment]
@yssource The error is:

> error LNK2019: unresolved external symbol "private: static unsigned int hmdf::ThreadGranularity::num_of_threads_

The problem is that you trying to access a symbol which wasn't exported, but DataFrame project doesn't support __declspec(dllexport)/__declspec(import). Even if you force CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS=ON it won't work. It's an upstream limitation.

My suggestion: Drop shared option for Windows, request such support to the upstream

**Comment by uilianries** [comment]
For some reason, this project is forcing MD runtime, even when is configured to MT. I didn't any clue about.

```
conan create . dataframe/1.5.0@ -s compiler="Visual Studio" -s compiler.version=15 -s compiler.runtime=MT

...
 cl /c /IC:\Users\uilia\.conan\data\dataframe\1.5.0\_\_\build\7d0ab7bed042914e1f2bacde0cf7e4d3b1a535bf\source_subfolder\include /W3 /WX- /diagnostics:classic /MP12 /O2 /Ob2 /D WIN32 /D _WINDOWS /D NDEBUG /D "CMAKE_INTDIR=\"Release\"" /D _MBCS /Gm- /EHsc /MD /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /std:c++14 /Fo"DataFrame.dir\Release\\" /Fd"DataFrame.dir\Release\DataFrame.pdb" /Gd /TP /errorReport:queue C:\Users\uilia\.conan\data\dataframe\1.5.0\_\_\build\7d0ab7bed042914e1f2bacde0cf7e4d3b1a535bf\source_subfolder\src\Vectors\HeteroVector.cc C:\Users\uilia\.conan\data\dataframe\1.5.0\_\_\build\7d0ab7bed042914e1f2bacde0cf7e4d3b1a535bf\source_subfolder\src\Vectors\HeteroView.cc C:\Users\uilia\.conan\data\dataframe\1.5.0\_\_\build\7d0ab7bed042914e1f2bacde0cf7e4d3b1a535bf\source_subfolder\src\Vectors\HeteroPtrView.cc C:\Users\uilia\.conan\data\dataframe\1.5.0\_\_\build\7d0ab7bed042914e1f2bacde0cf7e4d3b1a535bf\source_subfolder\src\Utils\ThreadGranularity.cc C:\Users\uilia\.conan\data\dataframe\1.5.0\_\_\build\7d0ab7bed042914e1f2bacde0cf7e4d3b1a535bf\source_subfolder\src\Utils\DateTime.cc C:\Users\uilia\.conan\data\dataframe\1.5.0\_\_\build\7d0ab7bed042914e1f2bacde0cf7e4d3b1a535bf\source_subfolder\src\MMap\MMapBase.cc C:\Users\uilia\.conan\data\dataframe\1.5.0\_\_\build\7d0ab7bed042914e1f2bacde0cf7e4d3b1a535bf\source_subfolder\src\MMap\MMapFile.cc C:\Users\uilia\.conan\data\dataframe\1.5.0\_\_\build\7d0ab7bed042914e1f2bacde0cf7e4d3b1a535bf\source_subfolder\src\MMap\MMapSharedMem.cc
```

**Comment by yssource** [comment]
I think the commit [d060e05](https://github.com/hosseinmoein/DataFrame/commit/d060e05a6863e1d18ff876071798ad8d94cfada7) has resolved the dllexport problem. 
And, I will re-create the package when its next release comes out.

**Comment by yssource** [comment]
Close. Open a new PR [v1.7.0](https://github.com/conan-io/conan-center-index/pull/803)

**Pull request closed without merging**